### PR TITLE
Remove access to contiguous TL buffer from BoxEncoder, Resize, Shapes and Warp

### PR DESCRIPTION
**Pull request by klecki**
## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [x] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

#### Additional information
- Affected modules and functionalities:

- Key points relevant for the review:

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/pipeline/data/buffer.h [hunk 1]
```diff
@@ -218,11 +218,17 @@ class DLL_PUBLIC Buffer {
   }
 
   /**
-   * @brief Sets the type of allocation (pinned/non-pinned) for
-   * CPU buffers
+   * @brief Return true if there was data allocation
+   */
+  inline bool has_data_allocation() {
+    return !!data_;
```
Comment (awolant) [inline dali/pipeline/data/buffer.h:224]: Why `!!` is needed here? Wouldn't the result be the same without it?

**Review by awolant — APPROVED** [review]

File: dali/pipeline/data/buffer.h (hunk 1 above)
Comment (klecki) [reply dali/pipeline/data/buffer.h:224]: `data_` is shared_ptr and it has bool-conversion defined, but `return data_` doesn't allow for implicit conversion, so I use `!!` as the idiomatic and shortest notation.
Other alternatives would be `data_.get() != nullptr` or `static_cast<bool>(data_)` which do look like a mouthful.

File: dali/pipeline/data/buffer.h [hunk 2]
```diff
@@ -218,11 +218,17 @@ class DLL_PUBLIC Buffer {
   }
 
   /**
-   * @brief Sets the type of allocation (pinned/non-pinned) for
-   * CPU buffers
+   * @brief Return true if there was data allocation
+   */
+  inline bool has_data_allocation() {
```
Comment (mzient) [inline dali/pipeline/data/buffer.h:223]:
```suggestion
  inline bool has_data {
```
`allocation` is not that clear when data is _shared_, not allocated by this buffer.

File: dali/operators/ssd/box_encoder.cu [hunk 3]
```diff
@@ -139,12 +139,12 @@ __device__ void MatchBoxWithAnchors(
...
-__global__ void Encode(
-  const float4 *boxes_in, const int *labels_in, const int *offsets, const int anchors_count,
-  const float4 *anchors, const float criteria, float4 *boxes_out,  int *labels_out,
-  int *box_idx_buffer, float *box_iou_buffer, bool offset, const float* means,
-  const float* stds, float scale, const float4 *anchors_as_cwh) {
-  const int sample = blockIdx.x;
+__global__ void Encode(const BoxEncoderSampleDesc *samples, const int anchors_count,
+                       const float4 *anchors, const float criteria, int *box_idx_buffer,
+                       float *box_iou_buffer, bool offset, const float *means, const float *stds,
+                       float scale, const float4 *anchors_as_cwh) {
```
Comment (mzient) [inline dali/operators/ssd/box_encoder.cu:145]: Why, oh, why?

File: dali/operators/ssd/box_encoder.cu [hunk 4]
```diff
@@ -153,13 +153,12 @@ __global__ void Encode(
   __shared__ volatile int best_anchor_idx_tmp[BLOCK_SIZE];
   __shared__ volatile float best_anchor_iou_tmp[BLOCK_SIZE];
 
-  volatile int *best_box_idx = box_idx_buffer + sample * anchors_count;
-  volatile float *best_box_iou = box_iou_buffer + sample * anchors_count;
+  volatile int *best_box_idx = box_idx_buffer + sample_idx * anchors_count;
+  volatile float *best_box_iou = box_iou_buffer + sample_idx * anchors_count;
```
Comment (mzient) [inline dali/operators/ssd/box_encoder.cu:157]: Please get rid of this volatile.

File: dali/pipeline/data/buffer.h (hunk 1 above)
Comment (mzient) [reply dali/pipeline/data/buffer.h:224]: I'm fine with !!, but doesn't it have a comparison with nullptr_t, so you could write `data_ != nullptr`?

Comment (klecki) [reply dali/pipeline/data/buffer.h:229]: Yup, probably could also do that.

**Review by mzient — APPROVED** [review]

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Pull request merged**